pkg/controllers/infra/cluster_controller.go

```diff
@@ -89,16 +100,32 @@ func (r *ClusterController) Reconcile(ctx context.Context, req ctrl.Request) (_
 		}
 	}()
 
+	if err := r.reconcileSecretOwnerReference(ctx, cluster); err != nil {
+		cluster.Status.Phase = string(clusterv1alpha1.ClusterPhaseFailed)
+		l.Error(err, "reconvile secret failed, please check the cluster's configuration")
+		// mark cluster failed, and do not requeue
+		return ctrl.Result{}, nil
+	}
+
 	scope := scope.NewCluster(cluster)
 	provider, err := r.newProvider(scope)
 	if err != nil {
 		conditions.MarkFalse(cluster, clusterv1alpha1.ReadyCondition, clusterv1alpha1.ProviderInitializeFailedReason, capiv1.ConditionSeverityError, err.Error())
 		return ctrl.Result{RequeueAfter: r.RequeueAfter}, errors.Wrapf(err, "failed to create Cluster Provider %s/%s", cluster.Namespace, cluster.Name)
 	}
 
+	if err := provider.Precheck(ctx); err != nil {
+		conditions.MarkFalse(cluster, clusterv1alpha1.ReadyCondition, clusterv1alpha1.PrecheckFailedReason, capiv1.ConditionSeverityError, err.Error())
+		cluster.Status.Phase = string(clusterv1alpha1.ClusterPhaseFailed)
+		l.Error(err, "precheck failed, please check the cluster's credential")
```

Comment:
we should not mention cluster's credential here. It should be included in err msg

Reply:
should not assume credential file exist, it's optional.

Reply:
I mean you should mention credential in implementor. In the caller you should not assume what's wrong

Reply:
done